python/ray/data/_internal/execution/streaming_executor.py

```diff
@@ -44,6 +44,10 @@ def __init__(self, options: ExecutionOptions):
         self._global_info: Optional[ProgressBar] = None
         self._output_info: Optional[ProgressBar] = None
 
+        # The executor can be shutdown while still running.
+        self._shutdown_lock = threading.RLock()
```

Comment:
Hmm why does this need to be a reentrant lock? AFAICT any calls to `.shutdown()` should return (releasing the lock) before any other calls to `.shutdown()` from the same thread:
1. call from the `finally` block of `.execute()` happen at the end of the generator loop.
2. call from `Dataset.take()` would either call it after `.execute()` has completed (i.e. after the full `.shutdown()` from (1)), or when the generator has been cut short by the `limit` (where the call in (1) never happens).

Am I missing a case that requires reentrancy?

Reply:
It's just my habit to default to reentrant locks (similar to how Java synchronized blocks are reentrant)--- I don't see the use case for a non-reentrant lock here.